rasterio/_base.pyx

```diff
@@ -156,17 +161,24 @@ cdef class DatasetBase(object):
         log.debug("Dataset %r is started.", self)
 
     cdef GDALDatasetH handle(self) except NULL:
+        """Return the object's GDAL dataset handle"""
         return self._hds
 
     cdef GDALRasterBandH band(self, int bidx) except NULL:
-        cdef GDALRasterBandH hband = NULL
+        """Return a GDAL raster band handle"""
+        cdef GDALRasterBandH band = NULL
+
         try:
             with CPLErrors() as cple:
-                hband = _gdal.GDALGetRasterBand(self._hds, bidx)
+                band = _gdal.GDALGetRasterBand(self._hds, bidx)
                 cple.check()
-        except CPLE_IllegalArg as exc:
+        except CPLE_IllegalArgError as exc:
             raise IndexError(str(exc))
-        return hband
+
+        if band == NULL:
+            raise ValueError("NULL band")
```

Comment:
I removed some of these tests that weren't necessary, but a different message should be done in a different PR.